File: day16_state/guru/py/main.py
from context import Context
from state import *


def client_code() -> None:
    context = Context(ConcreteStateA())
    context.request_1()
    context.request_2()


if __name__ == '__main__':
    client_code()
    # Context: Transition to ConcreteStateA
    # ConcreteStateA handles request_1.
    # ConcreteStateA wants to change the state of the context.
    # Context: Transition to ConcreteStateB
    # ConcreteStateB handles request_2.
    # ConcreteStateB wants to change the state of the context.
    # Context: Transition to ConcreteStateA

File: day16_state/guru/cpp/main.cpp
#include <memory>

#include "context.hpp"
#include "state.hpp"


void ClientCode() {
  auto context = Context::CreateSharedWithInitState(std::make_shared<ConcreteStateA>());
  context->Request1();
  context->Request2();
}


int main(int argc, char const *argv[]) {
  ClientCode();
  // Context: Transition to ConcreteStateA.
  // ConcreteStateA handles Request1.
  // ConcreteStateA wants to change the state of the context.
  // Context: Transition to ConcreteStateB.
  // ConcreteStateB handles Request2.
  // ConcreteStateB wants to change the state of the context.
  // Context: Transition to ConcreteStateA.

  return 0;
}
